### PR TITLE
support python 3.5 in libs

### PR DESCRIPTION
**Pull request by dstathis**
Note: @balbirthomas I converted `JujuTopology` from a dataclass to a regular class. Tests still pass but I wanted to bring that to your attention.

### CODE REVIEW

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 1]
```diff
@@ -601,17 +605,16 @@ def _group_name(path) -> str:
...
 
         # Generate group name:
         #  - prefix, from the relative path of the rule file;
         #  - name, from juju topology
-        return (
-            f"{'' if relpath == '.' else relpath.replace(os.path.sep, '_') + '_'}"
-            f"{topology.identifier}_alerts"
+        return "{}{}_alerts".format(
+            "" if relpath == "." else relpath.replace(os.path.sep, "_") + "_", topology.identifier
         )
```
Comment (simskij) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:615]:
```suggestion
        prefix =  "" if relpath == "." else relpath.replace(os.path.sep, "_") + "_"
        return "{}{}_alerts".format(prefix, topology.identifier)
```

**Review by balbirthomas — APPROVED** [review]

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 2]
```diff
@@ -601,17 +605,16 @@ def _group_name(path) -> str:
...
 
-    for path in filter(Path.is_file, Path(dir_path).glob("**/*.rule" if recursive else "*.rule")):
+    for path in filter(Path.is_file, Path(dir_path).glob("**/*.rule" if recursive else "*.rule")):  # type: ignore
```
Comment (simskij) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:617]:
```suggestion
    paths = Path(dir_path).glob("**/*.rule" if recursive else "*.rule")
    for path in filter(Path.is_file, paths):  # type: ignore
```

**Review by simskij — DISMISSED** [review]
All in all, looks good, but I added a few suggestions that I think are sympathetic towards future readers.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 3]
```diff
@@ -476,14 +475,14 @@ def _sanitize_scrape_configuration(job) -> dict:
     return sanitized_job
 
 
-@dataclasses.dataclass(frozen=True)
 class JujuTopology:
     """Dataclass for storing and formatting juju topology information."""
```
Comment (sed-i) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:479]:
```suggestion
    """Class for storing and formatting juju topology information."""
```

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 4]
```diff
@@ -601,17 +605,16 @@ def _group_name(path) -> str:
...
+        relpath = os.path.relpath(os.path.dirname(path), str(dir_path))
 
         # Generate group name:
         #  - prefix, from the relative path of the rule file;
         #  - name, from juju topology
-        return (
-            f"{'' if relpath == '.' else relpath.replace(os.path.sep, '_') + '_'}"
-            f"{topology.identifier}_alerts"
+        return "{}{}_alerts".format(
+            "" if relpath == "." else relpath.replace(os.path.sep, "_") + "_", topology.identifier
```
Comment (sed-i) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:614]: While you're at it, @mmanciop correctly pointed out that it makes more sense to have the topology identifier first, and the rest to follow.

File: tox.ini [hunk 5]
```diff
@@ -52,6 +52,21 @@ commands =
...
+deps =
+    -r{toxinidir}/requirements.txt
+    mypy
+    types-dataclasses
+    types-PyYAML
+    #pytest
+    #pytest-operator
+    #juju
+    #types-setuptools
+    #types-toml
```
Comment (sed-i) [inline tox.ini:66]: To be removed?

File: tox.ini [hunk 6]
```diff
@@ -52,6 +52,21 @@ commands =
     isort --check-only --diff {[vars]all_path}
     black --check --diff {[vars]all_path}
 
+[testenv:static]
+description = Run static analysis checks
+deps =
+    -r{toxinidir}/requirements.txt
+    mypy
+    types-dataclasses
```
Comment (sed-i) [inline tox.ini:60]: To be removed?

**Review by sed-i — APPROVED** [review]

**Comment by sed-i** [comment]
Bump LIBPATCH?

File: tox.ini (hunk 5 above)
Comment (dstathis) [reply tox.ini:66]: yes

File: tox.ini (hunk 6 above)
Comment (dstathis) [reply tox.ini:60]: no

**Comment by dstathis** [comment]
manually merging so leon can continue work

**Pull request merged**